FixInvalidDayOffsetPreprocessor: Allow DURATION as value

sunkup opened the pull request:
Allow DURATION as value in FixInvalidDayOffsetPreprocessor so that it will be found and the actual duration value can be fixed if invalid.

rfc2822 reviewed (commented):
I assume this is based on a real issue? Which one?

Can we require a property (or a list of properties which may have a `DURATION`) or at least `;VALUE=`?

This pre-processor is a very risky and off-the-wall way to fix things and should be used in the most minimal way. For instance, I always fear that some plaintext may contain exactly this content and then be modified …

ArnyminerZ reviewed (approved):
Everything looks good, makes sense. I didn't know that there was such a property. Should we somehow use it at some point? I mean, we can only fetch new data if necessary knowing that it only updates once a day, for example.

sunkup commented:
Updated. 

I believe DURATION can only occur as value of the properties: DURATION, REFRESH-INTERVAL, RELATED-TO, TRIGGER. 

Since TRIGGER never occurs as a value of the list of properties (like DURATION does), it's a bit weird to have it in the regex as well, but I think separating it might make the regex unnecessarily complex.

sunkup commented on lib/src/main/kotlin/at/bitfire/ical4android/validation/FixInvalidDayOffsetPreprocessor.kt:
```diff
@@ -14,7 +14,9 @@ object FixInvalidDayOffsetPreprocessor : StreamPreprocessor() {
         // TRIGGER:-PT2D
-        "^(DURATION|TRIGGER):-?P((T-?\\d+D)|(-?\\d+DT))\$",
+        // REFRESH-INTERVAL;VALUE=DURATION:PT1D
+        "(?:^|(DURATION|REFRESH-INTERVAL|RELATED-TO|TRIGGER);VALUE=)" +
```
> What is the `|` at position 5 of the regex needed for? 

It ensures that the matched string starts with either 
- "DURATION;VALUE=" (or other property),  
- "DURATION:" or "TRIGGER:"

> Shouldn't it fail because xxDURATION is not in the list of processed properties?

No, we could solve that with another circumflex to denote line start, like this:

```
(?:^|(^(DURATION|REFRESH-INTERVAL|RELATED-TO|TRIGGER));VALUE=)(DURATION|TRIGGER):-?P((T-?\d+D)|(-?\d+DT))$
```

But then we can just used this shorter and more pleasant to look at version instead:

```
^((DURATION|REFRESH-INTERVAL|RELATED-TO|TRIGGER);VALUE=)|(DURATION|TRIGGER):-?P((T-?\d+D)|(-?\d+DT))$
```

*regex updated*

sunkup replied:
> But then we can just use this shorter and more pleasant to look at version instead:
> 
> ```
> ^((DURATION|REFRESH-INTERVAL|RELATED-TO|TRIGGER);VALUE=)|(DURATION|TRIGGER):-?P((T-?\d+D)|(-?\d+DT))$
> ```

I need to reverse on this one. The non-capturing group (`(?:X)`) is needed after all. It ensures "xxDURATION" or "FREE-BUSY" are ignored. So we should just stick with the bit more complex version I guess ... 

> ```
> (?:^|(^(DURATION|REFRESH-INTERVAL|RELATED-TO|TRIGGER));VALUE=)(DURATION|TRIGGER):-?P((T-?\d+D)|(-?\d+DT))$
> ```

-> _regex updated_

rfc2822 commented on lib/src/test/kotlin/at/bitfire/ical4android/validation/FixInvalidDayOffsetPreprocessorTest.kt:
```diff
@@ -27,6 +26,14 @@ class FixInvalidDayOffsetPreprocessorTest {
         assertEquals(expected, fixed)
     }
 
+    private fun fixAssertNotEqual(expected: String, testValue: String) {
+        // Fix the duration string
+        val fixed = FixInvalidDayOffsetPreprocessor.fixString(testValue)
+
+        // Assert
+        assertEquals(expected, fixed)
```
The method name contains `not equal`, but the assert is `assertEquals`, which is confusing. Shouldn't we call it `fixStringAndAssertEquals` (meaning that it first calls `fixString` and then `assertEquals`)?

sunkup replied:
No, I simply forgot to call `assertNotEquals()` for the assertion. Then the tests fail so we have to change the regex again. See https://github.com/bitfireAT/ical4android/pull/182#discussion_r1829015657

rfc2822 commented on lib/src/test/kotlin/at/bitfire/ical4android/validation/FixInvalidDayOffsetPreprocessorTest.kt:
```diff
@@ -12,8 +12,7 @@ import java.time.Duration
 
-    private fun fixAndAssert(expected: String, testValue: String) {
-
+    private fun fixParseAssertEqual(expected: String, testValue: String) {
```
This method can be used for all tests, true? At least when I replace every `fixAssertNotEqual` call by `fixParseAssertEqual`, all tests still succeed.

Can we have only one assert method for all cases? Otherwise we should find a better naming, and maybe also KDoc for the assert helpers so that it's clear what they do.

See also next comment.

rfc2822 reviewed (changes requested):
Working fine.

I have suggestions for method names in tests.

rfc2822 commented on lib/src/main/kotlin/at/bitfire/ical4android/validation/FixInvalidDayOffsetPreprocessor.kt:
```diff
@@ -14,7 +14,8 @@ object FixInvalidDayOffsetPreprocessor : StreamPreprocessor() {
         // TRIGGER:-PT2D
-        "^(DURATION|TRIGGER):-?P((T-?\\d+D)|(-?\\d+DT))\$",
+        // REFRESH-INTERVAL;VALUE=DURATION:PT1D
+        "(?:^|(^(DURATION|REFRESH-INTERVAL|RELATED-TO|TRIGGER));VALUE=)(DURATION|TRIGGER):-?P((T-?\\d+D)|(-?\\d+DT))\$",
```
Last question (I hope): I have tried without `?:` and it worked as well.

I think it doesn't really matter whether we use it, because the strings don't contain `PT` or `DT`.

But let's assume there would be a property `CAPTION` with potential `VALUE=DURATION` that should also be fixed – in this case it would rewrite the property name to `CATION`, right (because of `.replace("PT", "P")``?)

So it would probably make sense to mark _everything_ except the actual capturing group that should be rewritten as non-capturing?

sunkup replied:
> Last question (I hope): I have tried without `?:` and it worked as well.

How did you change the regex? Can you give a whole example? 

This: 
`"^|(^(DURATION|REFRESH-INTERVAL|RELATED-TO|TRIGGER));VALUE=(DURATION|TRIGGER):-?P((T-?\\d+D)|(-?\\d+DT))\$"`

fails for all tests with "Text cannot be parsed to a Duration", which makes sense I guess, since the whole string is not a valid definition of a duration.

rfc2822 replied:
Ok what I meant is, if we use this regex: `(?:^|(^(CAPTION|DURATION|REFRESH-INTERVAL|RELATED-TO|TRIGGER));VALUE=)(DURATION|TRIGGER):-?P((T-?\\d+D)|(-?\\d+DT))\$` – I have only added the made-up property CAPTION here, just for illustration of the hypothetical problem – we get:

result: `CAPION;VALUE=DURATION:P1D`  – note the `CAPION`

Which is unexpected because I just added CAPTION.

However we don't have to add CAPTION or any other property that contains `PT` or `DT`. 😄  So it seems fine.

---

The other thing is, I don't get the usage of the non-capturing group yet. This regex works here, too:

```
override fun regexpForProblem() = Regex(
        "(^|(^(DURATION|REFRESH-INTERVAL|RELATED-TO|TRIGGER));VALUE=)(DURATION|TRIGGER):-?P((T-?\\d+D)|(-?\\d+DT))\$",
        setOf(RegexOption.MULTILINE, RegexOption.IGNORE_CASE)
    )
```

Because we don't use the captured groups at all, but replace `DT` and `PT` in the full match (which is basically the core of the "hypothetical problem" mentioned above):

```
        for (match in found) {
            val matchStr = match.value   // we operate on the full match here, not on captured groups
            // for illustration, we can use this here (output is then shown in test logs):
            System.err.println("Replacing PT and DT in $matchStr (${match.groupValues.size} captured groups)")
            val fixed = matchStr
                .replace("PT", "P")
                .replace("DT", "D")
```

---

I also just noticed that `match` has a `match.range` and we can directly use it to call replace: `s.replace(match.range, fixed)`. If I understand it correctly, this should replace the actually found occurrence and not run a new search.

rfc2822 reviewed (commented):
I have changed the test so that it asserts what is expected (and not what is not expected, which was confusing for me).

And one question :)

rfc2822 reviewed (commented):
Sorry for making this complicated… but I think it's worth to fully understand the details of this regex and class and what they do, otherwise it could make unexpected problems.

sunkup commented:
> Sorry for making this complicated… but I think it's worth to fully understand the details of this regex and class and what they do, otherwise it could make unexpected problems.

No worries. I don't like working with regex, but you are totally right. We should not slip up here. I did not fully understand how the non-capturing groups work. I expected them to be excluded from the whole match, but alas they are not and we need to use the list of capturing groups like you say.

I changed the regex to capture only the duration in a capturing group, while ignoring the begining of the string, so problems as with the "CAPTION" scenario should not arise:
```
"(?:^|^(?:DURATION|REFRESH-INTERVAL|RELATED-TO|TRIGGER);VALUE=)(?:DURATION|TRIGGER):(-?P((T-?\\d+D)|(-?\\d+DT)))$"
```

I also tried to make the code as explicit as possible with many comments. The first capturing group always contains the whole match. The second match group (`match.groupValues[1]`) now only contains the actual faulty duration, ie "-PT1D", which is fixed first. Then the whole fixed duration string is replacing the faulty duration string in the whole match string and in the end the whole match string is found and replaced in the original ical.   

Best case we would probably to add another edge case test for the "CAPTION" scenario. If we are worried about unexpected issues, more tests clamping down on the expected behaviour are always nice to have. I don't see how we can do that though, since "CAPTION" is not a relevant property here as it can not have a duration value. 

> I also just noticed that match has a match.range and we can directly use it to call replace: s.replace(match.range, fixed). If I understand it correctly, this should replace the actually found occurrence and not run a new search.

AFAIK `replace()` does not take an int range, so I think we would need to use [`replaceRange()`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/replace-range.html) for that.  I don't think we need to use the match range at all though.

rfc2822 commented:
To illustrate what I mean: I have now used `replaceRange` so that we exactly replace the match, so we don't have to find multiple times (with potential different results).

Tests are OK; please have a final look and then I'll merge if it's OK for you.

sunkup commented:
> To illustrate what I mean: I have now used `replaceRange` so that we exactly replace the match, so we don't have to find multiple times (with potential different results).
> 
> Tests are OK; please have a final look and then I'll merge if it's OK for you.

Oh, so the match group has a range as well ... Yeah, no. I did not get that from your description. But looks good to me like this. Much shorter :+1:
